### ISSUE TITLE
Temporal activity heartbeat from incorrect thread

### ISSUE THREAD

**Issue by itaseskii** [airbytehq/airbyte#11515]
## Environment
- **Airbyte version**:  latest
- **OS Version / Instance**: Ubuntu 20
- **Deployment**: Docker
- **Source Connector and version**: N/A 
- **Destination Connector and version**: N/A 
- **Severity**: High
- **Step where error happened**: ReplicationActivityImpl, NormalizationActivityImpl, DbtTransformationActivityImpl 

## Current Behavior

The current **_TemporalUtils_** implementation when performing heartbeats to inform the **_Temporal_** server of the activity liveness it is doing so in a background thread scheduled by a **_ScheduledExecutorService_** which leads to two issues.
1. The **_ActivityContext_** itself is retrieved from a **_Thread_** which didn't start or is executing the activity, this will lead to an error since the Temporal SDK is retrieving the context from a **_ThreadLocal_** which in this case will be empty since you are making the call from an incorrect thread.

```
private static final ThreadLocal<ActivityExecutionContext> CURRENT = new ThreadLocal<>();

  /**
   * This is used by activity implementation to get access to the current ActivityExecutionContext
   */
  public static ActivityExecutionContext get() {
    ActivityExecutionContext result = CURRENT.get();
    if (result == null) {
      throw new IllegalStateException(
          "ActivityExecutionContext can be used only inside of activity "
              + "implementation methods and in the same thread that invoked an activity.");
    }
    return result;
  }
```

2. If the retrieval of the **_ActivityContext_** was to fail or a heartbeat couldn't be performed for some other reason the **_ScheduledExecutorService_** won't actually inform you of the error, it will fail silently and will lead to cancellation of any further work to be done by the runnable.

## Expected Behavior

Whenever using any **_Temporal SDK_** related classes that depend on a **_ThreadLocal_** implementation it is better if the scope of the call doesn't leave the thread on which the Workflow or Activity is executing or was being called on. For that reason I think that although a simple fix can be provided by extracting the **_Activity.getExecutionContext().heartbeat(null)_** call outside of the scheduler lambda and just passing the reference a line bellow it is worth investing a bit more effort in order to remove any such calls from classes like **_TemporalUtils_** because there is no guarantee in which thread they will be executed on. If I was to start a new Thread in the activity itself and invoke **_TemporalUtils.withBackgroundHeartbeat()_** I would encounter the same issue as it is now. This is not isolated to this case alone, the **_CancelationHandler_** will have the same issue if not called from the Thread running the activity.

## Logs

Activity execution stack trace:

<details>

<summary>LOG</summary> 

```

java.lang.IllegalStateException: ActivityExecutionContext can be used only inside of activity implementation methods and in the same thread that invoked an activity.
	at io.temporal.internal.activity.CurrentActivityExecutionContext.get(CurrentActivityExecutionContext.java:40)
	at io.temporal.internal.activity.ActivityInternal.getExecutionContext(ActivityInternal.java:29)
	at io.temporal.activity.Activity.getExecutionContext(Activity.java:40)
	at io.airbyte.workers.temporal.TemporalUtils.lambda$withBackgroundHeartbeat$3(TemporalUtils.java:226)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)


```

</details>

## Steps to Reproduce
You can use the following test code to reproduce the issue in your **_Temporal_** tests:
https://gist.github.com/itaseskii/da9585b9f4ca7d480e47f728d2a4326b. 
Just make sure to wrap the **_Activity.getExecutionContext().heartbeat(null);_** inside the **_ScheduledExecutorService.scheduleAtFixedRate()_** in a **_try catch_** block otherwise the **_ScheduledExecutorService_** will fail silently suspending any further work to be done.

## Are you willing to submit a PR?
I will be glad to provide a PR fixing this issue and even do some refactoring of the current implementation.

**Comment by jrhizor** [comment]
Thanks for the very clear issue + the test to reproduce! We'd be happy to review a fix. Please let us know if there's input you need from our side to do the refactor.

**Comment by itaseskii** [comment]
Great! There are no questions for now, but once I start working on the issue If something needs to be clarified related to the implementation details I will let you know,

**Comment by itaseskii** [comment]
Quick update. I have started working on this issue.

**Comment by itaseskii** [comment]
This will require a bit more effort and thinking things through. The **_LauncherWorker_** is calling **_TemporalUtils.withBackgroundHeartbeat_** from within its _**run**_ method and with the current implementation in order to avoid the context retrieval to escape the thread running the activity  I don't think there is any other way to pass the **_ActivityExecutionContext_** other than to extend the general **_Worker.run()_** method with an additional parameter which I don't think is an ideal solution since it will expose all **_Workers_** needlessly to a domain specific object in this case the **_ActivityExecutionContext_** from the Temporal SDK. @jrhizor What do you think?

**Comment by marcosmarxm** [comment]
Closed by #11878
